// marko/dist/runtime/vdom/AsyncVDOMBuilder.js
var EventEmitter = require("events-light");
var vdom = require("./vdom");
var VElement = vdom.aW_;
var VDocumentFragment = vdom.aX_;
var VComment = vdom.aY_;
var VText = vdom.aZ_;
var VComponent = vdom.b__;
var VFragment = vdom.ba_;
var virtualizeHTML = vdom.bb_;
var RenderResult = require("../RenderResult");
var defaultDocument = vdom.bc_;
var morphdom = require("./morphdom");
var attrsHelper = require("./helpers/attrs");

var EVENT_UPDATE = "update";
var EVENT_FINISH = "finish";

function State(tree) {
  this.bd_ = new EventEmitter();
  this.be_ = tree;
  this.bf_ = false;
}

function AsyncVDOMBuilder(globalData, parentNode, parentOut) {
  if (!parentNode) {
    parentNode = new VDocumentFragment();
  }

  var state;

  if (parentOut) {
    state = parentOut.M_;
  } else {
    state = new State(parentNode);
  }

  this.bg_ = 1;
  this.bh_ = 0;
  this.bi_ = null;
  this.bj_ = parentOut;

  this.data = {};
  this.M_ = state;
  this.aC_ = parentNode;
  this.global = globalData || {};
  this.bk_ = [parentNode];
  this.bl_ = false;
  this.bm_ = undefined;
  this.h_ = null;

  this.l_ = null;
  this.n_ = null;
  this.aA_ = null;
}

var proto = AsyncVDOMBuilder.prototype = {
  aS_: true,
  ___: defaultDocument,

  bc: function (component, key, ownerComponent) {
    var vComponent = new VComponent(component, key, ownerComponent);
    return this.bn_(vComponent, 0, true);
  },

  aD_: function (component, key, ownerComponent) {
    var vComponent = new VComponent(component, key, ownerComponent, true);
    this.bn_(vComponent, 0);
  },

  bn_: function (child, childCount, pushToStack) {
    this.aC_.bo_(child);
    if (pushToStack === true) {
      this.bk_.push(child);
      this.aC_ = child;
    }
    return childCount === 0 ? this : child;
  },

  element: function (tagName, attrs, key, component, childCount, flags, props) {
    var element = new VElement(tagName, attrs, key, component, childCount, flags, props);
    return this.bn_(element, childCount);
  },

  aR_: function (tagName, attrs, key, componentDef, props) {
    return this.element(tagName, attrsHelper(attrs), key, componentDef.m_, 0, 0, props);
  },

  n: function (node, component) {
    // NOTE: We do a shallow clone since we assume the node is being reused
    //       and a node can only have one parent node.
    var clone = node.__();
    this.node(clone);
    clone.aF_ = component;

    return this;
  },

  node: function (node) {
    this.aC_.bo_(node);
    return this;
  },

  text: function (text, ownerComponent) {
    var type = typeof text;

    if (type != "string") {
      if (text == null) {
        return;
      } else if (type === "object") {
        if (text.toHTML) {
          return this.h(text.toHTML(), ownerComponent);
        }
      }

      text = text.toString();
    }

    this.aC_.bo_(new VText(text, ownerComponent));
    return this;
  },

  comment: function (comment, ownerComponent) {
    return this.node(new VComment(comment, ownerComponent));
  },

  html: function (html, ownerComponent) {
    if (html != null) {
      var vdomNode = virtualizeHTML(html, this.___ || document, ownerComponent);
      this.node(vdomNode);
    }

    return this;
  },

  beginElement: function (tagName, attrs, key, component, childCount, flags, props) {
    var element = new VElement(tagName, attrs, key, component, childCount, flags, props);
    this.bn_(element, childCount, true);
    return this;
  },

  aP_: function (tagName, attrs, key, componentDef, props) {
    return this.beginElement(tagName, attrsHelper(attrs), key, componentDef.m_, 0, 0, props);
  },

  p_: function (key, component, preserve) {
    var fragment = new VFragment(key, component, preserve);
    this.bn_(fragment, null, true);
    return this;
  },

  q_: function () {
    this.endElement();
  },

  endElement: function () {
    var stack = this.bk_;
    stack.pop();
    this.aC_ = stack[stack.length - 1];
  },

  end: function () {
    this.aC_ = undefined;

    var remaining = --this.bg_;
    var parentOut = this.bj_;

    if (remaining === 0) {
      if (parentOut) {
        parentOut.bp_();
      } else {
        this.bq_();
      }
    } else if (remaining - this.bh_ === 0) {
      this.br_();
    }

    return this;
  },

  bp_: function () {
    var remaining = --this.bg_;

    if (remaining === 0) {
      var parentOut = this.bj_;
      if (parentOut) {
        parentOut.bp_();
      } else {
        this.bq_();
      }
    } else if (remaining - this.bh_ === 0) {
      this.br_();
    }
  },

  bq_: function () {
    var state = this.M_;
    state.bf_ = true;
    state.bd_.emit(EVENT_FINISH, this.aT_());
  },

  br_: function () {
    var lastArray = this._last;

    var i = 0;

    function next() {
      if (i === lastArray.length) {
        return;
      }
      var lastCallback = lastArray[i++];
      lastCallback(next);

      if (!lastCallback.length) {
        next();
      }
    }

    next();
  },

  error: function (e) {
    try {
      this.emit("error", e);
    } finally {
      // If there is no listener for the error event then it will
      // throw a new Error here. In order to ensure that the async fragment
      // is still properly ended we need to put the end() in a `finally`
      // block
      this.end();
    }

    return this;
  },

  beginAsync: function (options) {
    if (this.bl_) {
      throw Error("Tried to render async while in sync mode. Note: Client side await is not currently supported in re-renders (Issue: #942).");
    }

    var state = this.M_;

    if (options) {
      if (options.last) {
        this.bh_++;
      }
    }

    this.bg_++;

    var documentFragment = this.aC_.bs_();
    var asyncOut = new AsyncVDOMBuilder(this.global, documentFragment, this);

    state.bd_.emit("beginAsync", {
      out: asyncOut,
      parentOut: this
    });

    return asyncOut;
  },

  createOut: function () {
    return new AsyncVDOMBuilder(this.global);
  },

  flush: function () {
    var events = this.M_.bd_;

    if (events.listenerCount(EVENT_UPDATE)) {
      events.emit(EVENT_UPDATE, new RenderResult(this));
    }
  },

  E_: function () {
    return this.M_.be_;
  },

  aT_: function () {
    return this.bt_ || (this.bt_ = new RenderResult(this));
  },

  on: function (event, callback) {
    var state = this.M_;

    if (event === EVENT_FINISH && state.bf_) {
      callback(this.aT_());
    } else if (event === "last") {
      this.onLast(callback);
    } else {
      state.bd_.on(event, callback);
    }

    return this;
  },

  once: function (event, callback) {
    var state = this.M_;

    if (event === EVENT_FINISH && state.bf_) {
      callback(this.aT_());
    } else if (event === "last") {
      this.onLast(callback);
    } else {
      state.bd_.once(event, callback);
    }

    return this;
  },

  emit: function (type, arg) {
    var events = this.M_.bd_;
    switch (arguments.length) {
      case 1:
        events.emit(type);
        break;
      case 2:
        events.emit(type, arg);
        break;
      default:
        events.emit.apply(events, arguments);
        break;
    }
    return this;
  },

  removeListener: function () {
    var events = this.M_.bd_;
    events.removeListener.apply(events, arguments);
    return this;
  },

  sync: function () {
    this.bl_ = true;
  },

  isSync: function () {
    return this.bl_;
  },

  onLast: function (callback) {
    var lastArray = this._last;

    if (lastArray === undefined) {
      this._last = [callback];
    } else {
      lastArray.push(callback);
    }

    return this;
  },

  D_: function (doc) {
    var node = this.bm_;
    if (!node) {
      var vdomTree = this.E_();
      // Create the root document fragment node
      doc = doc || this.___ || document;
      this.bm_ = node = vdomTree.bu_(doc, null);
      morphdom(node, vdomTree, doc, this.h_);
    }
    return node;
  },

  toString: function (doc) {
    var docFragment = this.D_(doc);
    var html = "";

    var child = docFragment.firstChild;
    while (child) {
      var nextSibling = child.nextSibling;
      if (child.nodeType != 1) {
        var container = docFragment.ownerDocument.createElement("div");
        container.appendChild(child.cloneNode());
        html += container.innerHTML;
      } else {
        html += child.outerHTML;
      }

      child = nextSibling;
    }

    return html;
  },

  then: function (fn, fnErr) {
    var out = this;
    var promise = new Promise(function (resolve, reject) {
      out.on("error", reject).on(EVENT_FINISH, function (result) {
        resolve(result);
      });
    });

    return Promise.resolve(promise).then(fn, fnErr);
  },

  catch: function (fnErr) {
    return this.then(undefined, fnErr);
  },

  isVDOM: true,

  c: function (componentDef, key, customEvents) {
    this.l_ = componentDef;
    this.n_ = key;
    this.aA_ = customEvents;
  }
};

proto.e = proto.element;
proto.be = proto.beginElement;
proto.ee = proto.aQ_ = proto.endElement;
proto.t = proto.text;
proto.h = proto.w = proto.write = proto.html;

module.exports = AsyncVDOMBuilder;
