// marko/dist/runtime/vdom/VNode.js
/* jshint newcap:false */
function VNode() {}

VNode.prototype = {
  bv_: function (finalChildCount, ownerComponent) {
    this.bN_ = finalChildCount;
    this.bO_ = 0;
    this.bD_ = null;
    this.bP_ = null;
    this.bA_ = null;
    this.bB_ = null;
    this.aF_ = ownerComponent;
  },

  get a_() {
    var firstChild = this.bD_;

    if (firstChild && firstChild.bC_) {
      var nestedFirstChild = firstChild.a_;
      // The first child is a DocumentFragment node.
      // If the DocumentFragment node has a first child then we will return that.
      // Otherwise, the DocumentFragment node is not *really* the first child and
      // we need to skip to its next sibling
      return nestedFirstChild || firstChild.b_;
    }

    return firstChild;
  },

  get b_() {
    var nextSibling = this.bB_;

    if (nextSibling) {
      if (nextSibling.bC_) {
        var firstChild = nextSibling.a_;
        return firstChild || nextSibling.b_;
      }
    } else {
      var parentNode = this.bA_;
      if (parentNode && parentNode.bC_) {
        return parentNode.b_;
      }
    }

    return nextSibling;
  },

  bo_: function (child) {
    this.bO_++;

    if (this.bF_ === "textarea") {
      if (child.bQ_) {
        var childValue = child.bw_;
        this.bG_ = (this.bG_ || "") + childValue;
      } else if (child.bz_) {
        this.bR_ = true;
      } else {
        throw TypeError();
      }
    } else {
      var lastChild = this.bP_;

      child.bA_ = this;

      if (lastChild) {
        lastChild.bB_ = child;
      } else {
        this.bD_ = child;
      }

      this.bP_ = child;
    }

    return child;
  },

  bI_: function finishChild() {
    if (this.bO_ === this.bN_ && this.bA_) {
      return this.bA_.bI_();
    } else {
      return this;
    }
  }

  // ,toJSON: function() {
  //     var clone = Object.assign({
  //         nodeType: this.nodeType
  //     }, this);
  //
  //     for (var k in clone) {
  //         if (k.startsWith('_')) {
  //             delete clone[k];
  //         }
  //     }
  //     delete clone._nextSibling;
  //     delete clone._lastChild;
  //     delete clone.parentNode;
  //     return clone;
  // }
};

module.exports = VNode;
