// marko/dist/runtime/components/ComponentDef.js
"use strict";

var complain;
var componentUtil = require("./util");
var attachBubblingEvent = componentUtil._z_;
var addDelegatedEventHandler = require("./event-delegation")._A_;
var extend = require("raptor-util/extend");
var KeySequence = require("./KeySequence");

var FLAG_WILL_RERENDER_IN_BROWSER = 1;
// var FLAG_HAS_BODY_EL = 2;
// var FLAG_HAS_HEAD_EL = 4;
var FLAG_OLD_HYDRATE_NO_CREATE = 8;

/**
 * A ComponentDef is used to hold the metadata collected at runtime for
 * a single component and this information is used to instantiate the component
 * later (after the rendered HTML has been added to the DOM)
 */
function ComponentDef(component, componentId, componentsContext) {
  this._B_ = componentsContext; // The AsyncWriter that this component is associated with
  this.m_ = component;
  this.id = componentId;

  this._C_ = undefined; // An array of DOM events that need to be added (in sets of three)

  this._D_ = false;

  this._E_ = false;
  this._F_ = 0;

  this._G_ = 0; // The unique integer to use for the next scoped ID
  this._H_ = null;
}

ComponentDef.prototype = {
  _I_: function (key) {
    return (this._H_ || (this._H_ = new KeySequence()))._I_(key);
  },

  /**
   * This helper method generates a unique and fully qualified DOM element ID
   * that is unique within the scope of the current component.
   */
  elId: function (nestedId) {
    var id = this.id;

    if (nestedId == null) {
      return id;
    } else {
      if (typeof nestedId !== "string") {

        nestedId = String(nestedId);
        // eslint-disable-next-line no-constant-condition
      }

      if (nestedId.indexOf("#") === 0) {
        id = "#" + id;
        nestedId = nestedId.substring(1);
      }

      return id + "-" + nestedId;
    }
  },
  /**
   * Returns the next auto generated unique ID for a nested DOM element or nested DOM component
   */
  _J_: function () {
    return this.id + "-c" + this._G_++;
  },

  d: function (eventName, handlerMethodName, isOnce, extraArgs) {
    addDelegatedEventHandler(eventName);
    return attachBubblingEvent(this, handlerMethodName, isOnce, extraArgs);
  },

  get e_() {
    return this.m_.e_;
  }
};

ComponentDef.prototype.nk = ComponentDef.prototype._I_;

ComponentDef._K_ = function (o, types, global, registry) {
  var id = o[0];
  var typeName = types[o[1]];
  var input = o[2] || null;
  var extra = o[3];

  var isLegacy = extra.l;
  var state = extra.s;
  var componentProps = extra.w;
  var flags = extra.f;

  var component = typeName /* legacy */ && registry._L_(typeName, id, isLegacy);

  // Prevent newly created component from being queued for update since we area
  // just building it from the server info
  component.X_ = true;

  if (!isLegacy && flags & FLAG_WILL_RERENDER_IN_BROWSER && !(flags & FLAG_OLD_HYDRATE_NO_CREATE)) {
    if (component.onCreate) {
      component.onCreate(input, { global: global });
    }
    if (component.onInput) {
      input = component.onInput(input, { global: global }) || input;
    }
  } else {
    if (state) {
      var undefinedPropNames = extra.u;
      if (undefinedPropNames) {
        undefinedPropNames.forEach(function (undefinedPropName) {
          state[undefinedPropName] = undefined;
        });
      }
      // We go through the setter here so that we convert the state object
      // to an instance of `State`
      component.state = state;
    }

    if (componentProps) {
      extend(component, componentProps);
    }

    if (isLegacy) {
      component.widgetConfig = extra.c;
      component._M_ = extra.a;
    }
  }

  component.T_ = input;

  if (extra.b) {
    component.Q_ = extra.b;
  }

  var scope = extra.p;
  var customEvents = extra.e;
  if (customEvents) {
    component._w_(customEvents, scope);
  }

  component.V_ = global;

  return {
    id: id,
    m_: component,
    _N_: extra.r,
    _C_: extra.d,
    _F_: extra.f || 0
  };
};

module.exports = ComponentDef;
