// marko/dist/runtime/components/ComponentsContext.js
"use strict";

var GlobalComponentsContext = require("./GlobalComponentsContext");

function ComponentsContext(out, parentComponentsContext) {
  var globalComponentsContext;
  var componentDef;

  if (parentComponentsContext) {
    globalComponentsContext = parentComponentsContext.j_;
    componentDef = parentComponentsContext._O_;

    var nestedContextsForParent;
    if (!(nestedContextsForParent = parentComponentsContext._P_)) {
      nestedContextsForParent = parentComponentsContext._P_ = [];
    }

    nestedContextsForParent.push(this);
  } else {
    globalComponentsContext = out.global.h_;
    if (globalComponentsContext === undefined) {
      out.global.h_ = globalComponentsContext = new GlobalComponentsContext(out);
    }
  }

  this.j_ = globalComponentsContext;
  this.h_ = [];
  this.B_ = out;
  this._O_ = componentDef;
  this._P_ = undefined;
  this.s_ = parentComponentsContext && parentComponentsContext.s_;
}

ComponentsContext.prototype = {
  C_: function (doc) {
    var componentDefs = this.h_;

    ComponentsContext._Q_(componentDefs, doc);

    this.B_.emit("_R_");

    // Reset things stored in global since global is retained for
    // future renders
    this.B_.global.h_ = undefined;

    return componentDefs;
  }
};

function getComponentsContext(out) {
  return out.h_ || (out.h_ = new ComponentsContext(out));
}

module.exports = exports = ComponentsContext;

exports.r_ = getComponentsContext;
