// marko/dist/runtime/components/Component.js
"use strict";
/* jshint newcap:false */

var complain;

var domInsert = require("../dom-insert");
var defaultCreateOut = require("../createOut");
var getComponentsContext = require("./ComponentsContext").r_;
var componentsUtil = require("./util");
var componentLookup = componentsUtil.F_;
var emitLifecycleEvent = componentsUtil.G_;
var destroyNodeRecursive = componentsUtil.H_;
var EventEmitter = require("events-light");
var RenderResult = require("../RenderResult");
var SubscriptionTracker = require("listener-tracker");
var inherit = require("raptor-util/inherit");
var updateManager = require("./update-manager");
var morphdom = require("../vdom/morphdom");
var eventDelegation = require("./event-delegation");
var domData = require("./dom-data");
var componentsByDOMNode = domData.I_;
var keyedElementsByComponentId = domData.J_;
var CONTEXT_KEY = "__subtree_context__";

var slice = Array.prototype.slice;

var COMPONENT_SUBSCRIBE_TO_OPTIONS;
var NON_COMPONENT_SUBSCRIBE_TO_OPTIONS = {
  addDestroyListener: false
};

var emit = EventEmitter.prototype.emit;
var ELEMENT_NODE = 1;

function removeListener(removeEventListenerHandle) {
  removeEventListenerHandle();
}

function walkFragments(fragment) {
  var node;

  while (fragment) {
    node = fragment.firstChild;

    if (!node) {
      break;
    }

    fragment = node.fragment;
  }

  return node;
}

function handleCustomEventWithMethodListener(component, targetMethodName, args, extraArgs) {
  // Remove the "eventType" argument
  args.push(component);

  if (extraArgs) {
    args = extraArgs.concat(args);
  }

  var targetComponent = componentLookup[component.K_];
  var targetMethod = typeof targetMethodName === "function" ? targetMethodName : targetComponent[targetMethodName];
  if (!targetMethod) {
    throw Error("Method not found: " + targetMethodName);
  }

  targetMethod.apply(targetComponent, args);
}

function resolveKeyHelper(key, index) {
  return index ? key + "_" + index : key;
}

function resolveComponentIdHelper(component, key, index) {
  return component.id + "-" + resolveKeyHelper(key, index);
}

/**
 * This method is used to process "update_<stateName>" handler functions.
 * If all of the modified state properties have a user provided update handler
 * then a rerender will be bypassed and, instead, the DOM will be updated
 * looping over and invoking the custom update handlers.
 * @return {boolean} Returns true if if the DOM was updated. False, otherwise.
 */
function processUpdateHandlers(component, stateChanges, oldState) {
  var handlerMethod;
  var handlers;

  for (var propName in stateChanges) {
    if (stateChanges.hasOwnProperty(propName)) {
      var handlerMethodName = "update_" + propName;

      handlerMethod = component[handlerMethodName];
      if (handlerMethod) {
        (handlers || (handlers = [])).push([propName, handlerMethod]);
      } else {
        // This state change does not have a state handler so return false
        // to force a rerender
        return;
      }
    }
  }

  // If we got here then all of the changed state properties have
  // an update handler or there are no state properties that actually
  // changed.
  if (handlers) {
    // Otherwise, there are handlers for all of the changed properties
    // so apply the updates using those handlers

    handlers.forEach(function (handler) {
      var propertyName = handler[0];
      handlerMethod = handler[1];

      var newValue = stateChanges[propertyName];
      var oldValue = oldState[propertyName];
      handlerMethod.call(component, newValue, oldValue);
    });

    emitLifecycleEvent(component, "update");

    component.L_();
  }

  return true;
}

function checkInputChanged(existingComponent, oldInput, newInput) {
  if (oldInput != newInput) {
    if (oldInput == null || newInput == null) {
      return true;
    }

    var oldKeys = Object.keys(oldInput);
    var newKeys = Object.keys(newInput);
    var len = oldKeys.length;
    if (len !== newKeys.length) {
      return true;
    }

    for (var i = 0; i < len; i++) {
      var key = oldKeys[i];
      if (oldInput[key] !== newInput[key]) {
        return true;
      }
    }
  }

  return false;
}

var componentProto;

/**
 * Base component type.
 *
 * NOTE: Any methods that are prefixed with an underscore should be considered private!
 */
function Component(id) {
  EventEmitter.call(this);
  this.id = id;
  this.M_ = null;
  this.N_ = null;
  this.O_ = null;
  this.P_ = null;
  this.Q_ = null; // Used to keep track of bubbling DOM events for components rendered on the server
  this.R_ = null;
  this.K_ = null;
  this.S_ = null;
  this.T_ = undefined;
  this.U_ = false;
  this.V_ = undefined;
  this.W_ = false;
  this.X_ = false;
  this.Y_ = false;
  this.Z_ = false;
  this.___ = undefined;

  var ssrKeyedElements = keyedElementsByComponentId[id];

  if (ssrKeyedElements) {
    this.o_ = ssrKeyedElements;
    delete keyedElementsByComponentId[id];
  } else {
    this.o_ = {};
  }
}

Component.prototype = componentProto = {
  _a_: true,

  subscribeTo: function (target) {
    if (!target) {
      throw TypeError();
    }

    var subscriptions = this.O_ || (this.O_ = new SubscriptionTracker());

    var subscribeToOptions = target._a_ ? COMPONENT_SUBSCRIBE_TO_OPTIONS : NON_COMPONENT_SUBSCRIBE_TO_OPTIONS;

    return subscriptions.subscribeTo(target, subscribeToOptions);
  },

  emit: function (eventType) {
    var customEvents = this.R_;
    var target;

    if (customEvents && (target = customEvents[eventType])) {
      var targetMethodName = target[0];
      var isOnce = target[1];
      var extraArgs = target[2];
      var args = slice.call(arguments, 1);

      handleCustomEventWithMethodListener(this, targetMethodName, args, extraArgs);

      if (isOnce) {
        delete customEvents[eventType];
      }
    }

    if (this.listenerCount(eventType)) {
      return emit.apply(this, arguments);
    }
  },
  getElId: function (key, index) {
    if (!key) {
      return this.id;
    }
    return resolveComponentIdHelper(this, key, index);
  },
  getEl: function (key, index) {
    if (key) {
      var resolvedKey = resolveKeyHelper(key, index);
      var keyedElement = this.o_["@" + resolvedKey];

      if (!keyedElement) {
        var keyedComponentRoot = this.o_[resolvedKey];

        if (keyedComponentRoot) {

          return keyedComponentRoot.nodeType === 1 /** Node.ELEMENT_NODE */
          ? keyedComponentRoot : walkFragments(keyedComponentRoot);
          // eslint-disable-next-line no-constant-condition
        }
      }

      return keyedElement;
    } else {
      return this.el;
    }
  },
  getEls: function (key) {
    key = key + "[]";

    var els = [];
    var i = 0;
    var el;
    while (el = this.getEl(key, i)) {
      els.push(el);
      i++;
    }
    return els;
  },
  getComponent: function (key, index) {
    var rootNode = this.o_[resolveKeyHelper(key, index)];
    if (/\[\]$/.test(key)) {
      rootNode = rootNode && rootNode[Object.keys(rootNode)[0]];
      // eslint-disable-next-line no-constant-condition
    }
    return rootNode && componentsByDOMNode.get(rootNode);
  },
  getComponents: function (key) {
    var lookup = this.o_[key + "[]"];
    return lookup ? Object.keys(lookup).map(function (key) {
      return componentsByDOMNode.get(lookup[key]);
    }).filter(Boolean) : [];
  },
  destroy: function () {
    if (this.W_) {
      return;
    }

    var root = this.N_;

    this._b_();

    var nodes = root.nodes;

    nodes.forEach(function (node) {
      destroyNodeRecursive(node);

      if (eventDelegation._c_(node) !== false) {
        node.parentNode.removeChild(node);
      }
    });

    root.detached = true;

    delete componentLookup[this.id];
    this.o_ = {};
  },

  _b_: function () {
    if (this.W_) {
      return;
    }

    emitLifecycleEvent(this, "destroy");
    this.W_ = true;

    componentsByDOMNode.set(this.N_, undefined);

    this.N_ = null;

    // Unsubscribe from all DOM events
    this._d_();

    var subscriptions = this.O_;
    if (subscriptions) {
      subscriptions.removeAllListeners();
      this.O_ = null;
    }
  },

  isDestroyed: function () {
    return this.W_;
  },
  get state() {
    return this.M_;
  },
  set state(newState) {
    var state = this.M_;
    if (!state && !newState) {
      return;
    }

    if (!state) {
      state = this.M_ = new this._e_(this);
    }

    state._f_(newState || {});

    if (state.Y_) {
      this._g_();
    }

    if (!newState) {
      this.M_ = null;
    }
  },
  setState: function (name, value) {
    var state = this.M_;

    if (!state) {
      state = this.M_ = new this._e_(this);
    }
    if (typeof name == "object") {
      // Merge in the new state with the old state
      var newState = name;
      for (var k in newState) {
        if (newState.hasOwnProperty(k)) {
          state._h_(k, newState[k], true /* ensure:true */);
        }
      }
    } else {
      state._h_(name, value, true /* ensure:true */);
    }
  },

  setStateDirty: function (name, value) {
    var state = this.M_;

    if (arguments.length == 1) {
      value = state[name];
    }

    state._h_(name, value, true /* ensure:true */
    , true /* forceDirty:true */
    );
  },

  replaceState: function (newState) {
    this.M_._f_(newState);
  },

  get input() {
    return this.T_;
  },
  set input(newInput) {
    if (this.Z_) {
      this.T_ = newInput;
    } else {
      this._i_(newInput);
    }
  },

  _i_: function (newInput, onInput, out) {
    onInput = onInput || this.onInput;
    var updatedInput;

    var oldInput = this.T_;
    this.T_ = undefined;
    this._j_ = out && out[CONTEXT_KEY] || this._j_;

    if (onInput) {
      // We need to set a flag to preview `this.input = foo` inside
      // onInput causing infinite recursion
      this.Z_ = true;
      updatedInput = onInput.call(this, newInput || {}, out);
      this.Z_ = false;
    }

    newInput = this.S_ = updatedInput || newInput;

    if (this.Y_ = checkInputChanged(this, oldInput, newInput)) {
      this._g_();
    }

    if (this.T_ === undefined) {
      this.T_ = newInput;
      if (newInput && newInput.$global) {
        this.V_ = newInput.$global;
      }
    }

    return newInput;
  },

  forceUpdate: function () {
    this.Y_ = true;
    this._g_();
  },

  _g_: function () {
    if (!this.X_) {
      this.X_ = true;
      updateManager._k_(this);
    }
  },

  update: function () {
    if (this.W_ === true || this._l_ === false) {
      return;
    }

    var input = this.T_;
    var state = this.M_;

    if (this.Y_ === false && state !== null && state.Y_ === true) {
      if (processUpdateHandlers(this, state._m_, state._n_, state)) {
        state.Y_ = false;
      }
    }

    if (this._l_ === true) {
      // The UI component is still dirty after process state handlers
      // then we should rerender

      if (this.shouldUpdate(input, state) !== false) {
        this._o_();
      }
    }

    this.L_();
  },

  get _l_() {
    return this.Y_ === true || this.M_ !== null && this.M_.Y_ === true;
  },

  L_: function () {
    this.Y_ = false;
    this.X_ = false;
    this.S_ = null;
    var state = this.M_;
    if (state) {
      state.L_();
    }
  },

  shouldUpdate: function () {
    return true;
  },

  G_: function (eventType, eventArg1, eventArg2) {
    emitLifecycleEvent(this, eventType, eventArg1, eventArg2);
  },

  _o_: function () {
    var self = this;
    var renderer = self._p_;

    if (!renderer) {
      throw TypeError();
    }

    var input = this.S_ || this.T_;

    updateManager._q_(function () {
      self._r_(input, false).afterInsert(self.___);
    });

    this.L_();
  },

  _r_: function (input, isHydrate) {
    var doc = this.___;
    var globalData = this.V_;
    var rootNode = this.N_;
    var renderer = this._p_;
    var createOut = renderer.createOut || defaultCreateOut;
    var out = createOut(globalData);
    out.sync();
    out.___ = this.___;
    out[CONTEXT_KEY] = this._j_;

    var componentsContext = getComponentsContext(out);
    var globalComponentsContext = componentsContext.j_;
    globalComponentsContext._s_ = this;
    globalComponentsContext.k_ = isHydrate;

    renderer(input, out);

    var result = new RenderResult(out);

    var targetNode = out.E_().a_;

    morphdom(rootNode, targetNode, doc, componentsContext);

    return result;
  },

  _t_: function () {
    var root = this.N_;
    root.remove();
    return root;
  },

  _d_: function () {
    var eventListenerHandles = this.P_;
    if (eventListenerHandles) {
      eventListenerHandles.forEach(removeListener);
      this.P_ = null;
    }
  },

  get _u_() {
    var state = this.M_;
    return state && state._v_;
  },

  _w_: function (customEvents, scope) {
    var finalCustomEvents = this.R_ = {};
    this.K_ = scope;

    customEvents.forEach(function (customEvent) {
      var eventType = customEvent[0];
      var targetMethodName = customEvent[1];
      var isOnce = customEvent[2];
      var extraArgs = customEvent[3];

      finalCustomEvents[eventType] = [targetMethodName, isOnce, extraArgs];
    });
  },

  get el() {
    return walkFragments(this.N_);
  },

  get els() {
    return (this.N_ ? this.N_.nodes : []).filter(function (el) {
      return el.nodeType === ELEMENT_NODE;
    });
    // eslint-disable-next-line no-constant-condition
  }
};

componentProto.elId = componentProto.getElId;
componentProto._x_ = componentProto.update;
componentProto._y_ = componentProto.destroy;

// Add all of the following DOM methods to Component.prototype:
// - appendTo(referenceEl)
// - replace(referenceEl)
// - replaceChildrenOf(referenceEl)
// - insertBefore(referenceEl)
// - insertAfter(referenceEl)
// - prependTo(referenceEl)
domInsert(componentProto, function getEl(component) {
  return component._t_();
}, function afterInsert(component) {
  return component;
});

inherit(Component, EventEmitter);

module.exports = Component;
