// marko/dist/runtime/components/State.js
var extend = require("raptor-util/extend");

function ensure(state, propertyName) {
  var proto = state.constructor.prototype;
  if (!(propertyName in proto)) {
    Object.defineProperty(proto, propertyName, {
      get: function () {
        return this._v_[propertyName];
      },
      set: function (value) {
        this._h_(propertyName, value, false /* ensure:false */);
      }
    });
  }
}

function State(component) {
  this.m_ = component;
  this._v_ = {};

  this.Y_ = false;
  this._n_ = null;
  this._m_ = null;
  this._X_ = null; // An object that we use to keep tracking of state properties that were forced to be dirty

  Object.seal(this);
}

State.prototype = {
  L_: function () {
    var self = this;

    self.Y_ = false;
    self._n_ = null;
    self._m_ = null;
    self._X_ = null;
  },

  _f_: function (newState) {
    var state = this;
    var key;

    var rawState = this._v_;

    for (key in rawState) {
      if (!(key in newState)) {
        state._h_(key, undefined, false /* ensure:false */
        , false /* forceDirty:false */
        );
      }
    }

    for (key in newState) {
      state._h_(key, newState[key], true /* ensure:true */
      , false /* forceDirty:false */
      );
    }
  },
  _h_: function (name, value, shouldEnsure, forceDirty) {
    var rawState = this._v_;

    if (shouldEnsure) {
      ensure(this, name);
    }

    if (forceDirty) {
      var forcedDirtyState = this._X_ || (this._X_ = {});
      forcedDirtyState[name] = true;
    } else if (rawState[name] === value) {
      return;
    }

    if (!this.Y_) {
      // This is the first time we are modifying the component state
      // so introduce some properties to do some tracking of
      // changes to the state
      this.Y_ = true; // Mark the component state as dirty (i.e. modified)
      this._n_ = rawState;
      this._v_ = rawState = extend({}, rawState);
      this._m_ = {};
      this.m_._g_();
    }

    this._m_[name] = value;

    if (value === undefined) {
      // Don't store state properties with an undefined or null value
      delete rawState[name];
    } else {
      // Otherwise, store the new value in the component state
      rawState[name] = value;
    }
  },
  toJSON: function () {
    return this._v_;
  }
};

module.exports = State;
